src/crimson/os/seastore/seastore_types.cc

```diff
@@ -239,20 +242,160 @@ std::string device_type_to_string(device_type_t dtype) {
+  dev_addr &= static_cast<common_addr_t>(
+    std::numeric_limits<device_segment_id_t>::max());
+  dev_addr |= static_cast<common_addr_t>(id & 0x8) << DEV_ADDR_LEN_BITS;
+  dev_addr |= static_cast<common_addr_t>(type)
+    << (std::numeric_limits<common_addr_t>::digits - 1);
+}
+
+#define PADDR_OPERATION(a_type, base, func)        \
+  if (get_addr_type() == a_type) {                 \
+    return static_cast<const base*>(this)->func;   \
+  }
+
+paddr_t paddr_t::operator-(paddr_t rhs) const {
+  if (get_addr_type() == addr_types_t::SEGMENT) {
+    return ((*static_cast<const seg_paddr_t*>(this)) - rhs);
+  }
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+bool paddr_t::is_relative() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_relative())
+  ceph_assert(0 == "not supported type");
+  return false;
+}
+
+bool paddr_t::is_record_relative() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_record_relative())
+  ceph_assert(0 == "not supported type");
+  return false;
+}
+
+paddr_t paddr_t::add_offset(int32_t o) const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, add_offset(o))
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+bool paddr_t::is_null() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_null())
+  ceph_assert(0 == "not supported type");
+  return true;
+}
+
+bool paddr_t::is_block_relative() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_block_relative())
+  ceph_assert(0 == "not supported type");
+  return false;
+}
+
+bool paddr_t::is_zero() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_zero())
+  ceph_assert(0 == "not supported type");
+  return true;
+}
+
+bool paddr_t::is_real() const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, is_real())
+  ceph_assert(0 == "not supported type");
+  return false;
+}
+
+paddr_t paddr_t::add_relative(paddr_t o) const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, add_relative(o))
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+paddr_t paddr_t::add_block_relative(paddr_t o) const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, add_block_relative(o))
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+paddr_t paddr_t::add_record_relative(paddr_t o) const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, add_record_relative(o))
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+paddr_t paddr_t::maybe_relative_to(paddr_t o) const {
+  PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, maybe_relative_to(o))
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
+#define PADDR_OPERATOR(a_type, base, oper, other)     \
+  if (get_addr_type() == a_type) {                    \
+    const base* l = static_cast<const base*>(this);   \
+    const base* r = static_cast<const base*>(&other); \
+    return *l oper *r;                                  \
+  }
+
```

Comment:
I don't think you need special handling here, we can just directly compare by paddr_t::dev_addr.  seg_paddr_t's behavior wouldn't change, and I wouldn't expect the random variant to either.